File: server/routes/ngo.js
const express = require('express');
const NGOModel = require('../models/ngo_models');
const NgoRouter = express.Router();
const User = require('../models/auth');
const TempNGOModel = require('../models/temp_ngo');
const  DonationModel = require('../models/donation');
NgoRouter.post("/api/addNgo",async(req,res)=>{
    try{
        const{id} = req.body;  
        console.log(req.body);
        let isExist = await TempNGOModel.findById(id);
        if(!isExist){
            return res.status(400).json({msg:"No Ngo is found!"});
        }
        await TempNGOModel.findByIdAndDelete(id);
        let ngo = new NGOModel({
            username:isExist.username,
            password:isExist.password,
            ngo_name:isExist.ngo_name,
            ngo_admin:isExist.ngo_admin,
            mobile_no:isExist.mobile_no,
            website:isExist.website,
            books:isExist.books,
            food:isExist.food,
            clothes:isExist.clothes,
            money:isExist.money,
            payNumber:isExist.payNumber,
            area:isExist.area,
            city:isExist.city,
            state:isExist.state,
            description:isExist.description,
            ngo_photo:isExist.ngo_photo,
            regNumber:isExist.regNumber,
            regProof:isExist.regProof,
            annualRepo:isExist.annualRepo,
        });
        ngo = await ngo.save();
        res.json(ngo);
    }
    catch(e){
        res.status(500).json({err:e.message});
    }
});

//decline Ngo
NgoRouter.post("/api/declineNgo", async(req,res) => {
    const {id} = req.body;
    await TempNGOModel.findByIdAndDelete(id);
    res.json({"msg":"Declined Ngo Successfully"});
})

//fetch ngo by city and donation
NgoRouter.get("/api/getNgoByCity",async(req,res)=>{
    try{
        let city = req.query.city;
        let type = req.query.type;
        let allNgoByCity = await NGOModel.find({city});
        if(allNgoByCity.length == 0){
            return res.status(400).json({msg:"No NGO found for this city!"});
        }
        let list=[];
        for(let i=0;i<allNgoByCity.length;i++){
            switch(type){
                case "books": if(allNgoByCity[i].books==1){
                            list.push(allNgoByCity[i]);
                        }
                break;
                case "food": if(allNgoByCity[i].food==1){
                    list.push(allNgoByCity[i]);
                }
                break;
                case "clothes": if(allNgoByCity[i].clothes==1){
                    list.push(allNgoByCity[i]);
                }
                break;
                case "money": if(allNgoByCity[i].money==1){
                    list.push(allNgoByCity[i]);
                }
                break;
            }
        }
        // console.log("This is new list:"+list);
        res.json(list);
    }
     
    catch(e){
        res.status(500).json({err:e.message});
    }
    //
});

//get ngo details
NgoRouter.get("/api/getNgoDetails",async(req,res)=>{
    try{
        let id = req.query.id;
        let findNgo = await NGOModel.findById(id);
        if(!findNgo){
            return res.status(400).json({msg:"No NGO found for this name!"});
        }
        res.json(findNgo);
    }
    catch(e){
        res.status(500).json({err:e.message});
    }
});


//addTemp ngo
NgoRouter.post("/api/addNgoTemp",async(req,res)=>{
    try{
        const{ngo_name,ngo_admin,mobile_no,website,books,food,clothes,money,payNumber,area,city,state,description,ngo_photo,regNumber,regProof,annualRepo} = req.body;  
        console.log(req.body);
        let isExist = await TempNGOModel.findOne({ngo_name});
        if(isExist){
            return res.status(400).json({msg:"Ngo is already added!"});
        }
        let ngo = new TempNGOModel({
            username:ngo_name,
            password:mobile_no,
            ngo_name,
            ngo_admin,
            mobile_no,
            website,
            books,
            food,
            clothes,
            money,
            payNumber,
            area,
            city,
            state,
            description,
            ngo_photo,
            regNumber,
            regProof,
            annualRepo,
        });
        ngo = await ngo.save();
        res.json(ngo);
    }
    catch(e){
        res.status(500).json({err:e.message});
    }
});


NgoRouter.get("/api/getAllTempNgo",async (req,res) => {
    try{
        let tempNgo = await TempNGOModel.find({});
        res.json(tempNgo);
    }catch(e){
        res.status(500).json({err:e.message})
    }
})

NgoRouter.get("/api/getAllNgo",async(req,res) => {
    try{
        let allNgos = await NGOModel.find({});
    res.json(allNgos);
    }
    catch(e){
        res.status(500).json({err:e.message});
    }

    //delete ngo

    NgoRouter.post("/api/deleteNgo",async(req,res) => {
        try{
            const {id}=req.body;
               
        let ngomodel = await NGOModel.findByIdAndDelete(id);
           ngomodel = await ngomodel.save();
           return ngomodel;
            
        }catch(e){
            res.status(500).json({err:e.message});
        }
    })});

    NgoRouter.post('/api/acceptDonation',async(req,res) => {
        try{
            const {donationId,userId}= req.body;
            let donation = await DonationModel.findById(donationId);
            if(donation.status != 0){
                return res.status(400).json({msg:'Donation is already Visited'})
            }
            donation = await DonationModel.findByIdAndUpdate(donationId,{status:1},{new:true});
            let user = await User.findById(userId);
            await User.findByIdAndUpdate(userId,{accepted:user.accepted+1});
            res.json({msg:"Donation accepted"})
        }catch(e){
            res.status(500).json({error:e.message})
            console.log(e.message);
        }
    });


    NgoRouter.post('/api/declineDonation',async(req,res) => {
        try{
            const {donationId,userId}=req.body;
            console.log(req.body);
            let donation = await DonationModel.findByIdAndUpdate(donationId,{status:2},{new:true});
            let user = await User.findById(userId);
            console.log(user);
            await User.findByIdAndUpdate(userId,{declined:user.declined+1});
            //console.log(donation);
            res.json({msg:"Donation declined"})
        }catch(e){
            res.status(500).json({error:e.message})
        }
    });

    NgoRouter.get('/api/searchNgo',async(req,res)=>{
        try{
            const ngos = await NGOModel.find({
                ngo_name:{$regex: req.query.name, $options: 'i'},
            });
            res.json(ngos);
        }catch(e){
            res.status(500).json({error:e.message})
     
        }

    })

module.exports = NgoRouter;

File: lib/models/ngo.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'dart:convert';

class Ngo {
  final String id;
  final String username;
  final String password;
  final String ngo_name;
  final String ngo_admin;
  final String city;
  final String area;
  final String state;
  final String description;
  final String mobile_no;
  final String ngo_photo;
  final String website;
  final int books;
  final int food;
  final int clothes;
  final int money;
  final String payNumber;
  final String regNumber;
  final String regProof;
  final String annualRepo;
  Ngo({
    required this.id,
    required this.username,
    required this.password,
    required this.ngo_name,
    required this.ngo_admin,
    required this.city,
    required this.area,
    required this.state,
    required this.description,
    required this.mobile_no,
    required this.ngo_photo,
    required this.website,
    this.books = 0,
    this.food = 0,
    this.clothes = 0,
    this.money = 0,
    required this.payNumber,
    required this.regNumber,
    required this.regProof,
    required this.annualRepo,
  });

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'username': username,
      'password': password,
      'id': id,
      'ngo_name': ngo_name,
      'ngo_admin': ngo_admin,
      'city': city,
      'area': area,
      'description': description,
      'mobile_no': mobile_no,
      'ngo_photo': ngo_photo,
      'website': website,
      'books': books,
      'food': food,
      'clothes': clothes,
      'money': money,
      'state':state,
      'payNumber': payNumber,
      'regNumber': regNumber,
      'regProof': regProof,
      'annualRepo': annualRepo,
    };
  }

  factory Ngo.fromMap(Map<String, dynamic> map) {
    return Ngo(
      state: map['state']??'',
      id: map['_id'] ?? "",
      ngo_name: map['ngo_name'] ?? " ",
      ngo_admin: map['ngo_admin'] ?? " ",
      city: map['city'] ?? " ",
      area: map['area'] ?? " ",
      description: map['description'] ?? " ",
      mobile_no: map['mobile_no'] ?? " ",
      ngo_photo: map['ngo_photo'] ?? " ",
      website: map['website'] ?? "",
      books: map['books'] ?? 0,
      food: map['food'] ?? 0,
      clothes: map['clothes'] ?? 0,
      money: map['money'] ?? 0,
      payNumber: map['payNumber'] ?? "",
      regNumber: map['regNumber'] ?? "",
      regProof: map['regProof'] ?? "",
      annualRepo: map['annualRepo'] ?? "",
      password: map['password'] ?? "",
      username: map['username'] ?? "",
    );
  }

  String toJson() => json.encode(toMap());

  factory Ngo.fromJson(String source) =>
      Ngo.fromMap(json.decode(source) as Map<String, dynamic>);
}
